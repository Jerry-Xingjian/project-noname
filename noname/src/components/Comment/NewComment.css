.new-comment-bar .form-width {
    width: 90% !important;
}

.new-comment-bar .commentInput{
    height: 60px;
    width: 90% !important;
}

.new-comment-bar .commentInput__input{
    height: 60px;
    max-width: 100% !important;
    border-radius: 5px;
    border-color: transparent !important;
    padding: 5px 10px !important;
    resize: none;
    overflow: scroll !important;
}

.new-comment-bar .commentInput__highlighter {
    height: 60px;
    padding: 5px 10px !important;
}

.new-comment-bar .mentionText {
    color: #6c757d;
    z-index: 1;
    position: relative;
}

.new-comment-bar .mention-suggestion .avatar-image {
    width: 24px;
    height: 24px;
    border-radius: 50%;
    margin-right: 10px;
}