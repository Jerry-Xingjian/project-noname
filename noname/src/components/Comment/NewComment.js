/* eslint-disable no-undef */
/* eslint-disable react/jsx-no-bind */
/* eslint-disable no-underscore-dangle */
import React, { useRef, useState } from 'react';
import { MentionsInput, Mention } from 'react-mentions';
import './NewComment.css';
import data from '../../utils/api/index';
import { addComment } from '../../utils/api/comment';

function NewComment(props) {
  const [comment, setComment] = useState('');
  const commentObj = useRef({});
  const {
    userInfo,
    postInfo,
    setCommentsState,
    showAlert,
  } = props;

  // Async function to query user from database
  // Should limit the number of users returned to 5
  function fetchUsers(query, callback) {
    if (!query) return;
    fetch(`${data.rootTemp}/search/users?username=${query}`, {
      json: true,
    })
      .then((res) => res.json())
      // Transform the users to what react-mentions expects
      .then((res) => res.data.map((user) => ({ ...user, display: `@${user.username}`, id: user._id })))
      .then(callback);
  }

  // Set the comment content value with mention whne comment change
  const handleCommentChange = async (e, _, newPlainTextValue, mentions) => {
    // Handle the changes in the textArea
    setComment(e.target.value);
    commentObj.current = {
      plainTextValue: newPlainTextValue,
      mentions,
    };
  };

  // Post new comment
  const handleNewComment = async (event) => {
    event.preventDefault();
    // Alert if the comment is empty
    if (comment === '' || comment.trim() === '') {
      showAlert('Comment content cannot be empty, please write a comment. ✍️');
      return;
    }
    // Clear comment box
    setComment('');
    // Init new comment object
    const postData = {
      belongPostId: postInfo._id,
      belongUserId: userInfo._id,
      content: commentObj.current,
    };

    // Make API call to add comment, then update comments state
    const res = await addComment(postData);
    let newComments = res.data.response.comments;
    // Sort the comments by createdAt
    newComments = newComments.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));
    setCommentsState(newComments);
  };

  // Function for a customeized mention dropdown list
  const renderSuggestion = (entry) => (
    <div className="d-flex px-1 py-1 mx-1 my-1 align-items-center mention-suggestion rounded">
      <img className="avatar-image me-3" src={entry.profilePicture} alt="Avatar" />
      <span>{entry.username}</span>
    </div>
  );

  return (
    <div className="new-comment-bar px-4 py-2 d-flex align-items-center">
      <img className="avatar-image me-3" src={userInfo.profilePicture} alt="Avatar" />
      <form className="d-flex form-width justify-content-between w-100" onSubmit={handleNewComment}>
        <MentionsInput
          id="floatingTextarea"
          className="commentInput w-100"
          data-testid="comment-textarea"
          name="comment"
          value={comment}
          onChange={handleCommentChange}
          placeholder="Leave a comment here"
        >
          <Mention
            trigger="@"
            // data={users}
            data={fetchUsers}
            renderSuggestion={renderSuggestion}
            className="mentionText"
          />
        </MentionsInput>
        <button data-testid="new-comment-post" type="submit" className="btn btn-link text-decoration-none">Post</button>
      </form>
    </div>
  );
}

export default NewComment;
