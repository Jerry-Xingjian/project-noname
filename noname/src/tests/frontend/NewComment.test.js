import React from "react";
import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import NewComment from '../../components/Comment/NewComment';

const defaultProp = {
  userInfo : {
    "_id": "6388320305d016d46329331d",
    "email": "qhao@seas.upenn.edu",
    "password": "$2b$10$Mxx3KERe2hI9GUEkb8Hsm.jr2gLNJyZYfooiGYK1RW0QBPdComYSO",
    "username": "qhao",
    "profilePicture": "https://nonome-project-media.s3.amazonaws.com/default_avatar.png",
    "followings": [],
    "followers": [],
    "createdAt": "2022-12-01T04:48:03.298Z",
    "updateAt": "2022-12-01T04:48:03.298Z",
    "bio": "This is some bio of the user qhao.\nThis can be super long but it's fine."
  },
  postInfo: {
    "_id": "638baf0616bf8bf4ec672316",
    "belongUserId": "6388320305d016d46329331d",
    "likeInfo": [],
    "media": [
        {
            "url": "https://nonome-project-media.s3.amazonaws.com/2.jpeg",
            "type": "image"
        }
    ],
    "caption": "1",
    "comments": [],
    "createdAt": "2022-12-03T20:18:14.794Z",
    "updateAt": "2022-12-03T20:18:14.794Z",
    "location": "2"
  },
  setCommentsState: jest.fn(),
}

describe("snapshot test", () => {
  test('newComments matches snapshot', () => {
    const component = renderer.create(<NewComment {...defaultProp} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
})

describe("Existence test", () => {
  test('renders new comment textarea', async() => {
    render(<NewComment {...defaultProp} />);
    expect(screen.getByText(/Post/i)).toBeInTheDocument();
  });
})

describe("Event test", () => {

  test("New comment onsubmit event test", async () => {
    render(<NewComment {...defaultProp}/>);
    const commentText = screen.getByTestId("comment-textarea");
    fireEvent.change(commentText, {target: {value: 'great post'}});
    await waitFor(() => {
      const button = screen.getByTestId("new-comment-post");
      fireEvent.click(button)
    })
    expect(screen.getByTestId("comment-textarea")).toHaveValue('');
  });

  test("changing comments test", async () => {
    render(<NewComment {...defaultProp}/>);
    await waitFor(() => {
      const textarea = screen.getByTestId("comment-textarea");
      expect(textarea).toHaveValue('');
      fireEvent.change(textarea, { target: { value: 'Test comment'} });
      expect(textarea.value).toBe('Test comment');
    })
  });
});