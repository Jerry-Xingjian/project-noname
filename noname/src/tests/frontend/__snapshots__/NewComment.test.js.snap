// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test newComments matches snapshot 1`] = `
<div
  className="new-comment-bar px-4 py-2 d-flex align-items-center"
>
  <img
    alt="Avatar"
    className="avatar-image me-3"
    src="https://nonome-project-media.s3.amazonaws.com/default_avatar.png"
  />
  <form
    className="d-flex form-width justify-content-between w-100"
    onSubmit={[Function]}
  >
    <div
      className="commentInput w-100 commentInput--multiLine"
      style={
        Object {
          "overflowY": "visible",
          "position": "relative",
        }
      }
    >
      <div
        className="commentInput__control"
        style={Object {}}
      >
        <div
          className="commentInput__highlighter"
          style={
            Object {
              "border": "1px solid transparent",
              "boxSizing": "border-box",
              "color": "transparent",
              "overflow": "hidden",
              "position": "relative",
              "textAlign": "start",
              "whiteSpace": "pre-wrap",
              "width": "100%",
              "wordWrap": "break-word",
            }
          }
        >
           
        </div>
        <textarea
          className="commentInput__input"
          data-testid="comment-textarea"
          id="floatingTextarea"
          name="comment"
          onBlur={[Function]}
          onChange={[Function]}
          onCompositionEnd={[Function]}
          onCompositionStart={[Function]}
          onKeyDown={[Function]}
          onScroll={[Function]}
          onSelect={[Function]}
          placeholder="Leave a comment here"
          style={
            Object {
              "backgroundColor": "transparent",
              "bottom": 0,
              "boxSizing": "border-box",
              "display": "block",
              "fontFamily": "inherit",
              "fontSize": "inherit",
              "height": "100%",
              "left": 0,
              "letterSpacing": "inherit",
              "margin": 0,
              "overflow": "hidden",
              "position": "absolute",
              "resize": "none",
              "top": 0,
              "width": "100%",
            }
          }
          value=""
        />
      </div>
    </div>
    <button
      className="btn btn-link text-decoration-none"
      data-testid="new-comment-post"
      type="submit"
    >
      Post
    </button>
  </form>
</div>
`;
